usr/src/cmd/svc/milestone/fs-joyent

```diff
@@ -87,11 +103,17 @@ if [ $? -ne 0 ]; then
 
     for pool in $pools; do
         zpool import -f $pool
-	# Due to early, failed attempts to support the filesystem_limits
-	# feature we now need to ensure the dependent feature is enabled.
-	zpool set feature@extensible_dataset=enabled $pool
+
+        is_encr="$(zfs get -Hpo value encryption $pool)"
+
+        [[ "$is_encr" != "off" ]] && unlock_pool $pool
```

Comment:
Later, we check for the `destroy_zpools` bootparam.  Maybe that should be moved up so that we don't bother trying to unlock a pool we will just destroy, particularly since it seems likely we are destroying the pool because we can't unlock it.

Reply:
A good idea.. I'll update.

Reply:
Actually thinking about it more.. we might want to leave this as-is.  The zfs encryption work added a `zpool import -l` option which loads all of the keys of the dataset. For 'prompt' and 'file://' values, it will load accordingly.  One bit of discussion that's been had among the OpenZFS developers is making it easier for each OS to supply their own hook for things like 'https://...' as a keylocation.  If that ever happens (I might try to do some work here as an EDAR follow on), something like a key location of 'ebox' or 'ebox://...' we could then just `zpool import -l $pool` here and have it DTRT (as well as hook in our libcurl here to support various methods as well). In that instance, we'd keep the same order of operations.

... though it may not end up mattering either way.. so if you'd still like it moved, just let me know and I'll adjust the code accordingly..

Reply:
If we have a boot option that says "destroy the pool[s]" it makes no sense to go through the effort/pain of trying to unlock a pool that will be unconditionally destroyed a millisecond after an unlock attempt completes (whether successful or not).

I am a fan of having better integration along the lines of what you suggest, but I don't see how that relates to this problem.  Even if the options for keysource expand, we will still need to have the SmartOS special for destroying zpools and will need to decide whether we should first have to unlock dataset(s) prior to destroying the pool they live in.

Reply:
That only works when the `destroy_zpools` boot parameter is set. The other method (used by `sdc-factoryreset`) sets `smartdc:factoryreset=true` on `${SYS_ZPOOL}/var`.  While we don't need the key loaded on a dataset to read properties, we do need to unlock a pool in order to check if `.system_pool` exists in the root dataset to identify the system pool before checking the properties of any <pool>/var dataset. 

As far as I can tell, there is also no tooling that actually uses the `destroy_zpools` boot parameter -- it appears to only be something an operator must explicitly provide during boot (I suppose they could also manually add it to booter as well). I also can't find any references in our documentation to it.  From what I and Trent can tell, it's an almost 10 year old option that's likely never used. The tooling all uses `sdc-factoryreset` (which sets the property on $SYS_ZPOOL/var).

I wouldn't argue with using a zfs property to indicate the system pool (which would make it possible to avoid loading the key when destroying the zpool), though I think that should probably be it's own change.

Reply:
Fair enough.  For Linux CNs I've been having the same thought about system pool property (`com.joyent:system_pool=true`).